### ISSUE TITLE
Exception when using AES GCM on Android API 19

### ISSUE THREAD

**Issue by mattmook** [google/tink#396]
Currently encrypting on Android API 19 using `AesGcmKeyManager.aes256GcmTemplate()` results in an `InvalidAlgorithmParameterException`:

```
java.security.InvalidAlgorithmParameterException: unknown parameter type.
    at com.android.org.bouncycastle.jcajce.provider.symmetric.util.BaseBlockCipher.engineInit(BaseBlockCipher.java:559)
    at javax.crypto.Cipher.init(Cipher.java:616)
    at javax.crypto.Cipher.init(Cipher.java:566)
    at com.google.crypto.tink.subtle.AesGcmJce.encrypt(AesGcmJce.java:75)
    at com.google.crypto.tink.aead.AeadWrapper$WrappedAead.encrypt(AeadWrapper.java:52)
    at com.appmattus.layercache.Encryption.encryptValue(Encryption.kt:87)
    at com.appmattus.layercache.Encryption.access$encryptValue(Encryption.kt:47)
```

This fails on API 19 because `AesGcmJce.getParams` will generate a `GCMParameterSpec` as the class exists, however, the cipher doesn't support it. Instead the logic in getParams on Android should use `IvParameterSpec` when `Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT` even though `GCMParameterSpec` appears "available".

Additionally, once the above is resolved, another error will be returned instead due to the call(s) to `updateAAD`:

```
java.lang.UnsupportedOperationException: This cipher does not support Authenticated Encryption with Additional Data
    at javax.crypto.CipherSpi.engineUpdateAAD(CipherSpi.java:393)
    at javax.crypto.Cipher.updateAAD(Cipher.java:1056)
    at com.google.crypto.tink.subtle.AesGcmJce.encrypt(AesGcmJce.java:77)
    at com.google.crypto.tink.aead.AeadWrapper$WrappedAead.encrypt(AeadWrapper.java:52)
    at com.appmattus.layercache.Encryption.encryptValue(Encryption.kt:87)
    at com.appmattus.layercache.Encryption.access$encryptValue(Encryption.kt:47)
```

The only way to resolve this is, again, specifically on API 19 ensuring that the `associatedData` to `Aead.encrypt` and `Aead.decrypt` is always null. Would it be better for the associatedData to not be used when running API 19 (or less) rather than relying on the user stumbling upon the comment in the source code and needing to guard against it themselves in their call to `encrypt`/`decrypt`?

Once those two issues are resolved then AES GCM encryption works on API 19.

**Comment by thaidn** [comment]
Good catch! I'll try to reproduce and fix it.

**Comment by thaidn** [comment]
>This fails on API 19 because AesGcmJce.getParams will generate a GCMParameterSpec as the class exists, however, the cipher doesn't support it. Instead the logic in getParams on Android should use IvParameterSpec when Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT even though GCMParameterSpec appears "available".

I'm fixing this. The change will be included in 1.5.0.

>The only way to resolve this is, again, specifically on API 19 ensuring that the associatedData to Aead.encrypt and Aead.decrypt is always null. Would it be better for the associatedData to not be used when running API 19 (or less) rather than relying on the user stumbling upon the comment in the source code and needing to guard against it themselves in their call to encrypt/decrypt?

This is a known issue, see the last bullet in https://github.com/google/tink/blob/master/docs/KNOWN-ISSUES.md#android. If a user sets an associated data, we can't silently ignore it, but we have to throw an exception to notify the user that what they wanted to do is not supported by Android.
